lib/charms/sdcore_webui_k8s/v0/sdcore_config.py

```diff
@@ -0,0 +1,343 @@
+
+    on = SdcoreConfigRequirerCharmEvents()
+
+    def __init__(self, charm: CharmBase, relation_name: str):
+        """Init."""
+        super().__init__(charm, relation_name)
+        self.charm = charm
+        self.relation_name = relation_name
+        self.framework.observe(charm.on[relation_name].relation_changed, self._on_relation_changed)
+        self.framework.observe(charm.on[relation_name].relation_broken, self._on_relation_broken)
+
+    def _on_relation_changed(self, event: RelationChangedEvent) -> None:
+        """Handle relation changed event.
+
+        Args:
+            event (RelationChangedEvent): Juju event.
+
+        Returns:
+            None
+        """
+        if remote_app_relation_data := self._get_remote_app_relation_data(event.relation):
+            self.on.webui_url_available.emit(
+                webui_url=remote_app_relation_data["webui_url"],
+            )
+
+    def _on_relation_broken(self, event: RelationBrokenEvent) -> None:
+        """Handle the Sdcore config relation broken event.
+
+        Args:
+            event (RelationBrokenEvent): Juju event.
+        """
+        self.on.webui_broken.emit()
+
+    @property
+    def webui_url(self) -> Optional[str]:
+        """Return the address of the webui GRPC endpoint.
+
+        Returns:
+            str: Endpoint address.
+        """
+        if remote_app_relation_data := self._get_remote_app_relation_data():
+            return remote_app_relation_data.get("webui_url")
+        return None
+
+    def _get_remote_app_relation_data(
+        self, relation: Optional[Relation] = None
+    ) -> Optional[Dict[str, str]]:
+        """Get relation data for the remote application.
+
+        Args:
+            relation: Juju relation object (optional).
+
+        Returns:
+            Dict: Relation data for the remote application
+            or None if the relation data is invalid.
+        """
+        relation = relation or self.model.get_relation(self.relation_name)
+
+        if not relation:
+            logger.error("No relation: %s", self.relation_name)
+            return None
+
+        if not relation.app:
+            logger.warning("No remote application in relation: %s", self.relation_name)
+            return None
+
+        remote_app_relation_data = dict(relation.data[relation.app])
+
+        if not data_is_valid(remote_app_relation_data):
+            logger.error("Invalid relation data: %s", remote_app_relation_data)
+            return None
+
+        return remote_app_relation_data
+
+
+class SdcoreConfigProvides(Object):
+    """Class to be instantiated by the charm providing the SD-Core Webui URL."""
+
+    def __init__(self, charm: CharmBase, relation_name: str):
+        """Init."""
+        super().__init__(charm, relation_name)
+        self.relation_name = relation_name
+        self.charm = charm
+
+    def set_webui_url(self, webui_url: str, relation_id: int) -> None:
```

Comment:
Do we need this method? I don't think I see a use case for it. Typically, we'll want to update all relations if the webui URL changes, right?

Reply:
The modules UDR, UDM, NSSF, NRF, AUSF, SMF, PCF, AMF will be the requirer of this relation. If the Webui URL is updated, it should be updated in all those relations.

Reply:
Right. That's my point. Below there's a method called `set_webui_url_in_all_relations` which will likely be our go to method. Does it make sense to also have a method updating only specific relation?

Reply:
In the webui side, when a single `RelationJoinedEvent` is triggered for `sdcore_config` interface, we need to set the Webui URL for this single relation.

Reply:
Of course, that makes sense :) So what would be the use for `set_webui_url_in_all_relations` then? I would see it useful if we could change the webui URL, but as far as I know it's not the case. Will we ever need to update all relations?

Reply:
I would use the `set_webui_url_in_all_relations` at the end of main callback method which runs on events such as `pebble ready`, `config changed`, `update status` etc. This will update all the relations at once if they exist without spending additional effort to get the available relations and update data. I would use `set_webui_url` when `sdcore_config` relations joined for a single relation. Does it make sense ?

Reply:
If we remove `set_webui_url_in_all_relations` from library, we probably require this method's functionality in the charm. Here, I would prefer to implement it in the library instead of the the charm itself.

Reply:
I'd agree with all you wrote if the `webui` URL was a config; but it's not. It can't change. As long as the charm is running, the URL pushed to the relation data bag will always be valid. IMO it's enough to publish the URL whenever new requirer joins the relation. 
On the provider side, even if the unit does down and a new unit is spun up, `relation-joined` will be fired as soon as the new provider unit joins the party. In which case, we're also covered without using this method.

Reply:
You are right. We do not need to set Webui URL under _configure. method. The method is removed.

Reply:
After discussing in the daily, `set_webui_url_in_all_relations` method is added again as Webui URL will be set under `_configure`.